src/main/java/com/aws/greengrass/mqttclient/v5/Publish.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.aws.greengrass.mqttclient.v5;
+
+import lombok.Builder;
+import lombok.NonNull;
+import lombok.Value;
+
+import java.util.List;
+
+@SuppressWarnings({"PMD.ClassWithOnlyPrivateConstructorsShouldBeFinal", "PMD.ExcessiveParameterList"})
+@Value
+public class Publish {
+    @NonNull String topic;
+    @NonNull QOS qos;
+    /**
+     * Retain the message in the cloud MQTT broker (only last message with retain is actually kept).
+     * Subscribers will immediately receive the last retained message when they first subscribe.
+     */
+    boolean retain;
+    byte[] payload;
+    PayloadFormatIndicator payloadFormat;
+    Long messageExpiryIntervalSeconds;
+    String responseTopic;
+    byte[] correlationData;
+    String contentType;
+    List<UserProperty> userProperties;
+
+    // Readonly - Not used for Publish, only set for received messages
+    List<Long> subscriptionIdentifiers;
+
+    @Builder
+    protected Publish(String topic, QOS qos, boolean retain, byte[] payload, PayloadFormatIndicator payloadFormat,
+                      Long messageExpiryIntervalSeconds, String responseTopic, byte[] correlationData,
+                      String contentType, List<UserProperty> userProperties, List<Long> subscriptionIdentifiers) {
+        // Intern the string to deduplicate topic strings in memory
+        this.topic = topic.intern();
```

Comment:
Do we need this here? Or should we intern later at the spooler?

I'm assuming the builder methods will call this constructor, but it seems if we got rid of the string interning that we could just let lombok do all of this for us

Reply:
This is the same that we do in the existing publish request for mqtt 3. I think that encapsulating it here like this simplifies things. No one really needs to look at the constructor implementation, just using the builder will all work fine. Keeps the spooler logic somewhat simpler.

Reply:
This would apply not only to IPC, but to any usage of publish internally, so doing the interning here is still appropriate IMO.